sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/pointer/CommonPointerLibraries.java

```diff
@@ -116,6 +138,48 @@ static boolean isMemberModifiable(LLVMPointerImpl receiver, String ident,
         }
     }
 
+    @ExportMessage
+    static boolean isMemberInvocable(LLVMPointerImpl receiver, String ident) {
+        LLVMInteropType type = receiver.getExportType();
+        if (type instanceof LLVMInteropType.Clazz) {
+            LLVMInteropType.Clazz clazz = (LLVMInteropType.Clazz) type;
+            return clazz.findMethod(ident) != null;
+        }
+        return false;
+    }
+
+    @ExportMessage
+    static Object invokeMember(LLVMPointerImpl receiver, String member, Object[] arguments)
+                    throws UnsupportedMessageException, ArityException, UnknownIdentifierException, UnsupportedTypeException {
+        LLVMInteropType type = receiver.getExportType();
+        if (!(type instanceof LLVMInteropType.Clazz)) {
+            throw UnsupportedTypeException.create(new Object[]{receiver}, receiver + " cannot be casted to LLVMInteropType.Clazz");
+        }
+        // change from receiver.foo(arguments) to interopLibrary.execute(foo, [receiver+arguments])
+        Object[] newArguments = new Object[arguments.length + 1];
+        newArguments[0] = receiver;
+        for (int i = 0; i < arguments.length; i++) {
+            newArguments[i + 1] = arguments[i];
+        }
+        LLVMInteropType.Clazz clazz = (LLVMInteropType.Clazz) receiver.getExportType();
+
+        Method method = clazz.findMethod(member, newArguments);
+        if (method == null) {
+            throw UnknownIdentifierException.create(member);
+        }
+        LLVMFunction llvmFunction = LLVMLanguage.getContext().getGlobalScope().getFunction(method.getLinkageName());
```

Comment:
Also, e should probably do some caching here, since this is quite inefficient to do every time we call a method. If both `clazz` and `member`, and the argument array length are all the same, then `llvmFunction` should also be the same, so we can cache it. At compile time, we'd only have the comparisons, and then a direct call, instead of an expensive lookup and an indirect call.